src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,355 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "actions-server.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Actions;
+using namespace chip::app::Clusters::Actions::Attributes;
+using namespace chip::Protocols::InteractionModel;
+
+Instance Instance::instance;
+Instance * Instance::GetInstance()
+{
+    return &instance;
+}
+
+void Instance::OnStateChanged(EndpointId endpoint, uint16_t actionId, uint32_t invokeId, ActionStateEnum actionState)
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnStateChanged");
+
+    // Record StateChanged event
+    EventNumber eventNumber;
+    Events::StateChanged::Type event{ actionId, invokeId, actionState };
+
+    if (CHIP_NO_ERROR != LogEvent(event, endpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to record OnStateChanged event");
+    }
+}
+
+void Instance::OnActionFailed(EndpointId endpoint, uint16_t actionId, uint32_t invokeId, ActionStateEnum actionState,
+                              ActionErrorEnum actionError)
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnActionFailed");
+
+    // Record ActionFailed event
+    EventNumber eventNumber;
+    Events::ActionFailed::Type event{ actionId, invokeId, actionState, actionError };
+
+    if (CHIP_NO_ERROR != LogEvent(event, endpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to record OnActionFailed event");
+    }
+}
+
+CHIP_ERROR Instance::ReadActionListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (uint16_t i = 0; i <= kMaxActionList; i++)
+    {
+        ActionStructStorage action;
+
+        CHIP_ERROR err = GetInstance()->mDelegate->ReadActionAtIndex(i, action);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(encoder.Encode(action));
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::ReadEndpointListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
```

Comment:
I will stop commenting on this, but all these GetInstance() calls in member methods are wrong.

Reply:
Fixed it.